lib/spack/spack/spec.py

```diff
@@ -4845,26 +4845,50 @@ def reconstruct_virtuals_on_edges(spec):
         spec: spec on which we want to reconstruct virtuals
     """
     # Collect all possible virtuals
-    possible_virtuals = set()
-    for node in spec.traverse():
-        try:
-            possible_virtuals.update(
-                {x for x in node.package.dependencies if spack.repo.PATH.is_virtual(x)}
-            )
-        except Exception as e:
-            warnings.warn(f"cannot reconstruct virtual dependencies on package {node.name}: {e}")
+    virtuals_needed, virtuals_provided = {}, {}
+    for edge in spec.traverse_edges(cover="edges", root=False):
+        parent_key = edge.parent.dag_hash()
+        if parent_key not in virtuals_needed:
+            # Construct which virtuals are needed by parent
+            virtuals_needed[parent_key] = set()
+            try:
+                parent_pkg = edge.parent.package
+            except Exception as e:
+                warnings.warn(
+                    f"cannot reconstruct virtual dependencies on {edge.parent.name}: {e}"
+                )
+                continue
+
+            for name, when_deps in parent_pkg.dependencies_by_name(when=True).items():
+                if not spack.repo.PATH.is_virtual(name):
+                    continue
+
+                if any(edge.parent.satisfies(x) for x in when_deps):
+                    virtuals_needed[parent_key].add(name)
+
+        if not virtuals_needed[parent_key]:
             continue
 
-    # Assume all incoming edges to provider are marked with virtuals=
-    for vspec in possible_virtuals:
-        try:
-            provider = spec[vspec]
-        except KeyError:
-            # Virtual not in the DAG
+        child_key = edge.spec.dag_hash()
+        if child_key not in virtuals_provided:
+            virtuals_provided[child_key] = set()
```

Comment:
```suggestion
```

Reply:
I can remove this, but it is there to avoid continuing to hit a warning for packages that throw